fv/rule_metadata_test.go

```diff
@@ -0,0 +1,178 @@
+// +build fvtests
+
+// Copyright (c) 2019-2020 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fv_test
+
+import (
+	"math/rand"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"github.com/projectcalico/felix/fv/containers"
+	"github.com/projectcalico/felix/fv/infrastructure"
+	"github.com/projectcalico/felix/fv/utils"
+	"github.com/projectcalico/felix/fv/workload"
+	api "github.com/projectcalico/libcalico-go/lib/apis/v3"
+	client "github.com/projectcalico/libcalico-go/lib/clientv3"
+)
+
+var _ = Describe("Rule Metadata tests", func() {
+
+	var (
+		//etcd   *containers.Container
+		felix  *infrastructure.Felix
+		client client.Interface
+		etcd   *containers.Container
+		wl0    *workload.Workload
+		wl1    *workload.Workload
+	)
+
+	BeforeEach(func() {
+		felix, etcd, client = infrastructure.StartSingleNodeEtcdTopology(infrastructure.DefaultTopologyOptions())
+
+		wl0 = workload.Run(felix, "test0", "default", "10.65.0.1", "80", "tcp")
+		wl0.Configure(client)
+
+		wl1 = workload.Run(felix, "test1", "default", "10.65.0.2", "80", "tcp")
+		wl1.Configure(client)
+	})
+
+	AfterEach(func() {
+		wl0.Stop()
+		wl1.Stop()
+		felix.Stop()
+		etcd.Stop()
+	})
+
+	Context("With a GlobalNetworkPolicy with rule metadata", func() {
+
+		var gnp *api.GlobalNetworkPolicy
+
+		BeforeEach(func() {
+			gnp = api.NewGlobalNetworkPolicy()
+			gnp.Name = "rule-meta-test"
+			gnp.Spec.Selector = "all()"
+			gnp.Spec.Ingress = []api.Rule{{
+				Action:   api.Allow,
+				Metadata: &api.RuleMetadata{Annotations: map[string]string{"fvtest": "gnp"}},
+			}}
+			_, err := client.GlobalNetworkPolicies().Create(utils.Ctx, gnp, utils.NoOptions)
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		It("should add comments to iptables", func() {
+			Eventually(getIPTables(felix.Name)).Should(ContainSubstring("-m comment --comment \"fvtest=gnp\""))
```

Comment:
Out of interest, do we get a comment for each annotation, or is it all annotations in one iptables comment?

Reply:
If it's the latter, I guess the next question would be how complex annotation values are quoted so as not to be ambiguous...